src/Polysemy/Cont/Internal.hs

```diff
@@ -95,4 +103,155 @@ embedSem = liftSem . weave (pure ()) (pure . join) inspectSem
 {-# INLINE embedSem #-}
 
 newtype Ref m s a = Ref { runRef :: a -> m s }
+
+instance Contravariant (Ref m s) where
+  contramap f ref = Ref (runRef ref . f)
+
 newtype ExitRef m a = ExitRef { enterExit :: forall b. a -> m b }
+
+instance Contravariant (ExitRef m) where
+  contramap f ref = ExitRef $ \a -> enterExit ref (f a)
+
+data ViaFreshRef uniq a = ViaFreshRef { getBacktrackException :: a -> (uniq, Any) }
```

Comment:
What is this `Any`? Could we existentialize whatever it is instead of unsafecoercing?

Reply:
The `Any` corresponds to `a` when the `ViaFreshRef` was originally created (as opposed to a `contramap`ped `ViaFreshRef`). We need it because we can't scope over that type variable, because then we wouldn't be able to lower other effects through. This is also why `uniq` is so important here; we create an `uniq` when we create the original `ViaFreshRef`, and then when we receive a backtrack exception, we check if it's `uniq` matches the one we created. If they do, we know the `Any` is the original `a`, and is thus safe to `unsafeCoerce` to `a`.
Existentializing doesn't help here, because `Subst` places no constraints (like `Typeable`) upon the original `a`, so we'd have to resort to `unsafeCoerce` anyway.